Fix dbm problems on Ubuntu / Travis

https://travis-ci.org/NREL/conan-openstudio-ruby/jobs/567005211#L2409

```
ext/extinit.c updated
compiling ext/extinit.o
generating enc.mk
compiling enc/encinit.c
linking ruby
*** Following extensions are not compiled:
dbm:
	Could not be configured. It will not be installed.
	Check ext/dbm/mkmf.log for more details.
*** Fix the problems, then remove these directories and try again if you want.
```

Might need some system dependencies etc.

This points to an issue with either conan failing to link to the required dependency  [gdbm/1.18.1@bincrafters/stable]9https://bintray.com/bincrafters/public-conan/gdbm%3Abincrafters/1.18.1%3Astable), or the gdbm lib that conan installs does not have content to build dbm.a  

When I install `apt-get install libgdbm-dev` before running conan, conan successfully links to those static libs and can compile the ext/dbm.a. 

I extracted a snippet of the /ext/dbm/mkmf.log which shows the linker looking in system folders. 
Does conan automatically try to link to required conan dependency paths (e.g  /home/conan/.conan/data/gdbm/...) or does this need to explicity set somewhere in the configs? 

We can hack it and run `apt-get install libgdbm-dev` but that kinda defeats the purposes of using conan. 

--------------------

have_type: checking for DBM in ndbm.h... -------------------- no

"/usr/bin/gcc -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict   -c conftest.c"
conftest.c:3:10: fatal error: ndbm.h: No such file or directory
 #include <ndbm.h>
          ^~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <ndbm.h>
4: 
5: /*top*/
6: typedef DBM conftest_type;
7: int conftestval[sizeof(conftest_type)?1:-1];
/* end */

--------------------

have_type: checking for DBM in qdbm/relic.h... -------------------- no

"/usr/bin/gcc -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict   -c conftest.c"
conftest.c:3:10: fatal error: qdbm/relic.h: No such file or directory
 #include <qdbm/relic.h>
          ^~~~~~~~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <qdbm/relic.h>
4: 
5: /*top*/
6: typedef DBM conftest_type;
7: int conftestval[sizeof(conftest_type)?1:-1];
/* end */

--------------------

have_type: checking for DBM in relic.h... -------------------- no

"/usr/bin/gcc -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict   -c conftest.c"
conftest.c:3:10: fatal error: relic.h: No such file or directory
 #include <relic.h>
          ^~~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <relic.h>
4: 
5: /*top*/
6: typedef DBM conftest_type;
7: int conftestval[sizeof(conftest_type)?1:-1];
/* end */

--------------------

----------------

have_library: checking for -lgdbm... -------------------- no

"/usr/bin/gcc -o conftest -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-const
ant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn 
-Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict conftest.c  -L. -L../.. -L. -fstack-protector -rdynamic -Wl,-export-dynamic     -Wl,-rpath,/home/conan/.conan/data/openstudio_ruby/2.5.5/nrel/stable/build/1814c0429e8b31430cfd31e305ad1ac9f11e
0703/Ruby-prefix/src/Ruby-install/lib -L/home/conan/.conan/data/openstudio_ruby/2.5.5/nrel/stable/build/1814c0429e8b31430cfd31e305ad1ac9f11e0703/Ruby-prefix/src/Ruby-install/lib -lruby-static -lgdbm  -lpthread -lgmp -ldl -lcrypt -lm   -lc"
/usr/bin/ld: cannot find -lgdbm
collect2: error: ld returned 1 exit status
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: /*top*/
 4: extern int t(void);
 5: int main(int argc, char **argv)
 6: {
 7:   if (argc > 1000000) {
 8:     printf("%p", &t);
 9:   }
10: 
11:   return 0;
12: }
13: 
14: int rb_hasFile(const char *t_filename) {
15:   return 0;
16: }
17: 
18: int rb_require_embedded(const char *t_filename) {
19:   return 0;
20: }
21: 
22: int t(void) { ; return 0; }
/* end */

This points to an issue with either conan failing to link to the required dependency  [gdbm/1.18.1@bincrafters/stable]9https://bintray.com/bincrafters/public-conan/gdbm%3Abincrafters/1.18.1%3Astable), or the gdbm lib that conan installs does not have content to build dbm.a  

When I install `apt-get install libgdbm-dev` before running conan, conan successfully links to those static libs and can compile the ext/dbm.a. 

I extracted a snippet of the /ext/dbm/mkmf.log which shows the linker looking in system folders. 
Does conan automatically try to link to required conan dependency paths (e.g  /home/conan/.conan/data/gdbm/...) or does this need to explicity set somewhere in the configs? 

We can hack it and run `apt-get install libgdbm-dev` but that kinda defeats the purposes of using conan. 

--------------------

have_type: checking for DBM in ndbm.h... -------------------- no

"/usr/bin/gcc -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict   -c conftest.c"
conftest.c:3:10: fatal error: ndbm.h: No such file or directory
 #include <ndbm.h>
          ^~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <ndbm.h>
4: 
5: /*top*/
6: typedef DBM conftest_type;
7: int conftestval[sizeof(conftest_type)?1:-1];
/* end */

--------------------

have_type: checking for DBM in qdbm/relic.h... -------------------- no

"/usr/bin/gcc -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict   -c conftest.c"
conftest.c:3:10: fatal error: qdbm/relic.h: No such file or directory
 #include <qdbm/relic.h>
          ^~~~~~~~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <qdbm/relic.h>
4: 
5: /*top*/
6: typedef DBM conftest_type;
7: int conftestval[sizeof(conftest_type)?1:-1];
/* end */

--------------------

have_type: checking for DBM in relic.h... -------------------- no

"/usr/bin/gcc -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn -Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict   -c conftest.c"
conftest.c:3:10: fatal error: relic.h: No such file or directory
 #include <relic.h>
          ^~~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <relic.h>
4: 
5: /*top*/
6: typedef DBM conftest_type;
7: int conftestval[sizeof(conftest_type)?1:-1];
/* end */

--------------------

----------------

have_library: checking for -lgdbm... -------------------- no

"/usr/bin/gcc -o conftest -I../../.ext/include/x86_64-linux -I../.././include -I../.././ext/dbm    -O3 -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-const
ant-logical-operand -Wno-self-assign -Wunused-variable -Wimplicit-int -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wmisleading-indentation -Wno-packed-bitfield-compat -Wsuggest-attribute=noreturn 
-Wsuggest-attribute=format -Wimplicit-fallthrough=0 -Wduplicated-cond -Wrestrict conftest.c  -L. -L../.. -L. -fstack-protector -rdynamic -Wl,-export-dynamic     -Wl,-rpath,/home/conan/.conan/data/openstudio_ruby/2.5.5/nrel/stable/build/1814c0429e8b31430cfd31e305ad1ac9f11e
0703/Ruby-prefix/src/Ruby-install/lib -L/home/conan/.conan/data/openstudio_ruby/2.5.5/nrel/stable/build/1814c0429e8b31430cfd31e305ad1ac9f11e0703/Ruby-prefix/src/Ruby-install/lib -lruby-static -lgdbm  -lpthread -lgmp -ldl -lcrypt -lm   -lc"
/usr/bin/ld: cannot find -lgdbm
collect2: error: ld returned 1 exit status
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: /*top*/
 4: extern int t(void);
 5: int main(int argc, char **argv)
 6: {
 7:   if (argc > 1000000) {
 8:     printf("%p", &t);
 9:   }
10: 
11:   return 0;
12: }
13: 
14: int rb_hasFile(const char *t_filename) {
15:   return 0;
16: }
17: 
18: int rb_require_embedded(const char *t_filename) {
19:   return 0;
20: }
21: 
22: int t(void) { ; return 0; }
/* end */

--- a/.ci/install.sh
+++ b/.ci/install.sh
@@ -25,9 +25,9 @@ if [[ $machine == Darwin ]]; then
   pyenv virtualenv 3.7.1 conan
   pyenv rehash
   pyenv activate conan
-elif [[ $machine == Linux ]]; then
-  # You're not root by default in conanio docker images
-  sudo apt install libgdbm-dev
+#elif [[ $machine == Linux ]]; then
+#  # You're not root by default in conanio docker images
+#  sudo apt install libgdbm-dev
 fi
 
 pip install conan --upgrade

--- a/.ci/install.sh
+++ b/.ci/install.sh
@@ -26,7 +26,8 @@ if [[ $machine == Darwin ]]; then
   pyenv rehash
   pyenv activate conan
 elif [[ $machine == Linux ]]; then
-  apt install libgdbm-dev
+  # You're not root by default in conanio docker images
+  sudo apt install libgdbm-dev
 fi
 
 pip install conan --upgrade

--- a/.ci/install.sh
+++ b/.ci/install.sh
@@ -2,22 +2,31 @@
 
 set -ex
 
-if [[ "$(uname -s)" == 'Darwin' ]]; then
-    brew update || brew update
-    brew outdated pyenv || brew upgrade pyenv
-    brew install pyenv-virtualenv
-    brew install cmake || true
-
-    if which pyenv > /dev/null; then
-        eval "$(pyenv init -)"
-    fi
-
-    pyenv install 3.7.1
-    pyenv virtualenv 3.7.1 conan
-    pyenv rehash
-    pyenv activate conan
+unameOut="$(uname -s)"
+case "${unameOut}" in
+  Linux*)     machine=Linux;;
+  Darwin*)    machine=Darwin;;
+  CYGWIN*|MINGW*|MSYS*)    machine=Windows;;
+  *)          machine="UNKNOWN:${unameOut}"
+esac
 
 
+if [[ $machine == Darwin ]]; then
+  brew update || brew update
+  brew outdated pyenv || brew upgrade pyenv
+  brew install pyenv-virtualenv
+  brew install cmake || true
+
+  if which pyenv > /dev/null; then
+    eval "$(pyenv init -)"
+  fi
+
+  pyenv install 3.7.1
+  pyenv virtualenv 3.7.1 conan
+  pyenv rehash
+  pyenv activate conan
+elif [[ $machine == Linux ]]; then
+  apt install libgdbm-dev
 fi
 
 pip install conan --upgrade

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -135,13 +135,14 @@ FindValue(CONAN_LIB_DIRS_GMP)
 # TODO: there might be side effects supplying "with-xxx-dir=''" if not defined. Problem is that generator expressions are evaluated as build time... So need to wrap into another
 # Note JM 2019-06-20: Please mind the space after the true_string arg.
 # Notice that we pass --with-dbm-dir poiting to conan gdbm: that's because we build with gdbm_compat which should provide dbm too
+# We also pass `--with-dbm-type=gdbm_compat` since we use GDBM > 1.8.1 (found this arg in `Ruby/ext/extconf.rb`)
 set(CONF_ARGS
 "$<IF:$<BOOL:${CURRENT_CONAN_ZLIB_ROOT}>,--with-zlib-dir=${CURRENT_CONAN_ZLIB_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_OPENSSL_ROOT}>,--with-openssl-dir=${CURRENT_CONAN_OPENSSL_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIBFFI_ROOT}>,--with-libffi-dir=${CURRENT_CONAN_LIBFFI_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIBYAML_ROOT}>,--with-libyaml-dir=${CURRENT_CONAN_LIBYAML_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_GDBM_ROOT}>,--with-gdbm-dir=${CURRENT_CONAN_GDBM_ROOT} ,>\
-$<IF:$<BOOL:${CURRENT_CONAN_GDBM_ROOT}>,--with-dbm-dir=${CURRENT_CONAN_GDBM_ROOT} ,>\
+$<IF:$<BOOL:${CURRENT_CONAN_GDBM_ROOT}>,--with-dbm-dir=${CURRENT_CONAN_GDBM_ROOT} --with-dbm-type=gdbm_compat ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_READLINE_ROOT}>,--with-readline-dir=${CURRENT_CONAN_READLINE_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_GMP_ROOT}>,--with-gmp-dir=${CURRENT_CONAN_GMP_ROOT} ,>"
 )

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -134,22 +134,26 @@ FindValue(CONAN_LIB_DIRS_GMP)
 
 # TODO: there might be side effects supplying "with-xxx-dir=''" if not defined. Problem is that generator expressions are evaluated as build time... So need to wrap into another
 # Note JM 2019-06-20: Please mind the space after the true_string arg.
+# Notice that we pass --with-dbm-dir poiting to conan gdbm: that's because we build with gdbm_compat which should provide dbm too
 set(CONF_ARGS
 "$<IF:$<BOOL:${CURRENT_CONAN_ZLIB_ROOT}>,--with-zlib-dir=${CURRENT_CONAN_ZLIB_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_OPENSSL_ROOT}>,--with-openssl-dir=${CURRENT_CONAN_OPENSSL_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIBFFI_ROOT}>,--with-libffi-dir=${CURRENT_CONAN_LIBFFI_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIBYAML_ROOT}>,--with-libyaml-dir=${CURRENT_CONAN_LIBYAML_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_GDBM_ROOT}>,--with-gdbm-dir=${CURRENT_CONAN_GDBM_ROOT} ,>\
+$<IF:$<BOOL:${CURRENT_CONAN_GDBM_ROOT}>,--with-dbm-dir=${CURRENT_CONAN_GDBM_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_READLINE_ROOT}>,--with-readline-dir=${CURRENT_CONAN_READLINE_ROOT} ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_GMP_ROOT}>,--with-gmp-dir=${CURRENT_CONAN_GMP_ROOT} ,>"
 )
 
+# TODO: once/if gdbm works on MSVC, check the gdbm_compat.lib name and others
 set(EXTLIBS
 "$<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_ZLIB}>,${CURRENT_CONAN_LIB_DIRS_ZLIB}/zlib.lib ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_OPENSSL}>,${CURRENT_CONAN_LIB_DIRS_OPENSSL}/libcrypto.lib ${CURRENT_CONAN_LIB_DIRS_OPENSSL}/libssl.lib ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_LIBFFI}>,${CURRENT_CONAN_LIB_DIRS_LIBFFI}/libffi.lib ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_LIBYAML}>,${CURRENT_CONAN_LIB_DIRS_LIBYAML}/yaml.lib ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_GDBM}>,${CURRENT_CONAN_LIB_DIRS_GDBM}/gdbm.lib ,>\
+$<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_GDBM}>,${CURRENT_CONAN_LIB_DIRS_GDBM}/gdbm_compat.lib ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_READLINE}>,${CURRENT_CONAN_LIB_DIRS_READLINE}/readline.lib ,>\
 $<IF:$<BOOL:${CURRENT_CONAN_LIB_DIRS_GMP}>,${CURRENT_CONAN_LIB_DIRS_GMP}/libgmp.lib ,> "
 )

--- a/conanfile.py
+++ b/conanfile.py
@@ -321,9 +321,10 @@ class OpenstudiorubyConan(ConanFile):
                               'usr-x64-mswin64_140.lib',
                               'wait_for_single_fd-x64-mswin64_140.lib']
 
+        n_libs = len(libs)
         n_expected_libs = len(expected_libs)
-        if (len(libs) == len(n_expected_libs)):
-            self.output.success("Found {} libs".format(len(libs)))
+        if (n_libs == n_expected_libs):
+            self.output.success("Found {} libs".format(n_libs))
 
         else:
             missing_libs = set(expected_libs) - set(libs)
@@ -337,7 +338,7 @@ class OpenstudiorubyConan(ConanFile):
                                   "{}".format(len(extra_libs), extra_libs))
 
             self.output.error("Found {} libs, expected {} "
-                              "libs".format(len(libs), len(n_expected_libs)))
+                              "libs".format(n_libs, n_expected_libs))
 
         # self.cpp_info.libdirs = ['lib', 'lib/ext', 'lib/enc']
         # Equivalent automatic detection

--- a/conanfile.py
+++ b/conanfile.py
@@ -235,7 +235,109 @@ class OpenstudiorubyConan(ConanFile):
         # Keep only the names:
         self.cpp_info.libs = [os.path.basename(x) for x in libs]
 
-        self.output.success("Found {} libs".format(len(libs)))
+        # These are the ext libs we expect on all platforms
+        ext_libs = [
+            'bigdecimal', 'bubblebabble', 'console', 'continuation',
+            'coverage', 'cparse', 'date_core', 'digest', 'escape', 'etc',
+            'fcntl', 'fiber', 'fiddle', 'generator', 'libenc', 'libtrans',
+            'md5', 'nkf', 'nonblock', 'objspace', 'openssl', 'parser',
+            'pathname', 'psych', 'ripper', 'rmd160', 'sdbm', 'sha1', 'sha2',
+            'sizeof', 'socket', 'stringio', 'strscan', 'wait', 'zlib']
+
+        if self.options.with_gdbm:
+            ext_libs += ['dbm', 'gdbm']
+        if self.options.with_readline:
+            ext_libs += ['readline']
+
+        # Not sure here...
+        if self.settings.os == 'Windows':
+            ext_libs += ['dltest', 'resolv']
+        elif self.settings.os in ['Linux', 'Macos']:
+            ext_libs += ['pty',
+                         'syslog']
+
+        # Now, append the lib extension, platform specific
+        ext_libs = ['{}.{}'.format(libname, libext) for libname in ext_libs]
+
+        if self.settings.os == 'Linux':
+            expected_libs = ['libruby-static.a'] + ext_libs
+
+        elif self.settings.os == 'Macos':
+            expected_libs = (['libruby.{v}-static.a'.format(v=self.version)] +
+                             ext_libs)
+
+        elif self.settings.os == 'Windows':
+            expected_libs = (['x64-vcruntime140-ruby250-static.lib'] +
+                             ext_libs)
+
+            expected_libs += ['at_exit-x64-mswin64_140.lib',
+                              'bignum-x64-mswin64_140.lib',
+                              'bug_3571-x64-mswin64_140.lib',
+                              'bug_5832-x64-mswin64_140.lib',
+                              'bug_reporter-x64-mswin64_140.lib',
+                              'call_without_gvl-x64-mswin64_140.lib',
+                              'class-x64-mswin64_140.lib',
+                              'compat-x64-mswin64_140.lib',
+                              'console-x64-mswin64_140.lib',
+                              'debug-x64-mswin64_140.lib',
+                              'dln-x64-mswin64_140.lib',
+                              'dot.dot-x64-mswin64_140.lib',
+                              'empty-x64-mswin64_140.lib',
+                              'exception-x64-mswin64_140.lib',
+                              'fd_setsize-x64-mswin64_140.lib',
+                              'file-x64-mswin64_140.lib',
+                              'float-x64-mswin64_140.lib',
+                              'foreach-x64-mswin64_140.lib',
+                              'funcall-x64-mswin64_140.lib',
+                              'hash-x64-mswin64_140.lib',
+                              'integer-x64-mswin64_140.lib',
+                              'internal_ivar-x64-mswin64_140.lib',
+                              'iseq_load-x64-mswin64_140.lib',
+                              'iter-x64-mswin64_140.lib',
+                              'memory_status-x64-mswin64_140.lib',
+                              'method-x64-mswin64_140.lib',
+                              'notimplement-x64-mswin64_140.lib',
+                              'num2int-x64-mswin64_140.lib',
+                              'numhash-x64-mswin64_140.lib',
+                              'path_to_class-x64-mswin64_140.lib',
+                              'postponed_job-x64-mswin64_140.lib',
+                              'printf-x64-mswin64_140.lib',
+                              'proc-x64-mswin64_140.lib',
+                              'protect-x64-mswin64_140.lib',
+                              'rational-x64-mswin64_140.lib',
+                              'rb_fatal-x64-mswin64_140.lib',
+                              'recursion-x64-mswin64_140.lib',
+                              'regexp-x64-mswin64_140.lib',
+                              'resize-x64-mswin64_140.lib',
+                              'scan_args-x64-mswin64_140.lib',
+                              'string-x64-mswin64_140.lib',
+                              'struct-x64-mswin64_140.lib',
+                              'symbol-x64-mswin64_140.lib',
+                              'thread_fd_close-x64-mswin64_140.lib',
+                              'time-x64-mswin64_140.lib',
+                              'tracepoint-x64-mswin64_140.lib',
+                              'typeddata-x64-mswin64_140.lib',
+                              'update-x64-mswin64_140.lib',
+                              'usr-x64-mswin64_140.lib',
+                              'wait_for_single_fd-x64-mswin64_140.lib']
+
+        n_expected_libs = len(expected_libs)
+        if (len(libs) == len(n_expected_libs)):
+            self.output.success("Found {} libs".format(len(libs)))
+
+        else:
+            missing_libs = set(expected_libs) - set(libs)
+            if missing_libs:
+                self.output.error("Missing {} libraries: "
+                                  "{}".format(len(missing_libs), missing_libs))
+
+            extra_libs = set(libs) - set(expected_libs)
+            if extra_libs:
+                self.output.error("Found {} extra libraries: "
+                                  "{}".format(len(extra_libs), extra_libs))
+
+            self.output.error("Found {} libs, expected {} "
+                              "libs".format(len(libs), len(n_expected_libs)))
 
         # self.cpp_info.libdirs = ['lib', 'lib/ext', 'lib/enc']
         # Equivalent automatic detection

